docs/naps/4-async-slicing.md

```diff
@@ -0,0 +1,636 @@
+
+    ...
+
+    def _slice_layers_async(self) -> None:
+        self._slicer.slice_layers_async(self.layers, self.dims)
+```
+
+The main response to the slice being ready occurs on the `QtViewer`
+because `QtViewer.layer_to_visual` provides a way to map from
+a layer to its corresponding vispy layer.
+
+```python
+class QtViewer:
+    ...
+
+    viewer: ViewerModel
+    layer_to_visual: Dict[Layer, VispyBaseLayer]
+
+    def __init__(self, ...):
+        ...
+        self.viewer._slicer.events.slice_ready.connect(self._on_slice_ready)
+
+    @ensure_main_thread
+    def _on_slice_ready(self, responses: ViewerSliceResponse):
+        for layer, response in responses.items():
+            if visual := self.layer_to_visual[layer]:
+                visual._set_slice(response)
+```
+
+The other reason is because `QtViewer` is a `QObject` that lives in the main
+thread, which makes it easier to ensure that the slice response is handled
+on the main thread.
+That's useful because Qt widgets and vispy nodes can only be safely updated
+on the main thread [^vispy-faq-threads], both of which occur when consuming
+slice output.
+
+
+## Implementation
+
+A tracking issue [^tracking-issue] serves as a way to communicate high level
+progress and discussion.
+
+A prototype [^prototype-pr] has been created that implements the approach described
+above to show its basic technical feasibility in napari's existing code base.
+It only implements it for some layer types, some slicing operations, and may
+not work when layer data is mutated.
+There are some other rough edges and plenty of ways to break it, but basic
+functionality should work.
+
+
+## Backward compatibility
+
+### Breaks synchronous slicing behavior
+
+The main goal of this project is to perform slicing asynchronously, so it's
+natural that we might break anyone that was depending on slicing being synchronous.
+At a minimum, we must provide a public way to achieve the same fundamental goals,
+such as connecting to the `slice_ready` event.
+
+### Store existing slice state on layer
+
+Many existing napari behaviors depend on the existing slice input and output state
+on the layer instances. In this proposal, we decide not to remove this state from the
+layer yet to prevent breaking other functionality that relies on it. As slice output is
+now generated asynchronously, we must ensure that this state is read and written atomically
+to mutually exclude the main and slicing thread from reading and writing inconsistent
+parts of that state.
+
+In order to do this, we plan to encapsulate the input and output state of each state into
+private dataclasses. There are no API changes, but this forces any read/write access of
+this state to acquire an associated lock.
+
+
+## Future work
+
+### Render each slice as soon as it is ready
+
+In this proposal, the slicing thread waits for slices of all layers to be ready before
+it emits the `slice_ready` event. There are a few reasons for that.
```

Comment:
I had another comment about this, but I see now you addressed it already. I think this would be a big improvement, especially because some slicing operations are CPU-intensive (`Points`) rather than I/O. But if it makes it easier to keep this for a later data, that's ok!

Reply:
Points slicing could be seedup (by the cost of memory), outside this NAP as we use suboptimal way of storage

Reply:
Yes, definitely! some thought already went into this from @kevinyamauchi :)